### ISSUE TITLE
MaybeEncodingError: Can't pickle large return objects?

### ISSUE THREAD

**Issue by nickeubank** [joblib/joblib#514]
My parallel jobs involve reading in big files (I have lots of memory, so running imports in parallel helps a lot on time), but I'm getting a 

    Files "/Users/Nick/anaconda/lib/python3.6/multiprocessing/pool.py", line 608 in get 
        raise self._value
    MaybeEncodingError: Error sending result: [my pandas dataframe as string]
    Reason: 'error("'i' format requires -2147483648 <= number <= 2147483647",)'

It appears to occur because the Parallel workers are returning an object that's too big to pickle. However, the [Python forums](https://bugs.python.org/issue17560) suggest that you can change the pickler easily in 3.5. 

Anyone know if there's a way to use that or any other work around?

**Comment by lesteve** [comment]
Closing as a duplicate of #387. Feel free to comment if you feel this is not the same problem.

Next time, can you paste the full stacktrace please? It may look very verbose but contains a lot of useful information.

**Comment by nickeubank** [comment]
@lesteve  This is similar to #387, but I'm wondering if the situation has changed. I notice there's a new bit of code in the joblib library that *appears* to allow for a custom pickler (so one could potentially use protocol 4 of cpickle)? https://github.com/joblib/joblib/blob/master/joblib/pool.py

Is that true, and if so, is that documented anywhere?

Also, here's a minimal replicating example [warning: my comp has 48 gb ram -- this may be a little too big for laptops. Haven't worked hard at the minimum Series size to generate the error]:

    import pandas as pd
    def prob_func(i):
	    return pd.Series(1).repeat(2**28)

    from joblib import Parallel, delayed
    a = Parallel(n_jobs=2, verbose=40)(delayed(prob_func)(i) for i in range(2))

Output:

    def prob_func(i):
        a = pd.Series(1).repeat(2**28)
        return a
    
    
    
    
    from joblib import Parallel, delayed                              
    a = Parallel(n_jobs=2, verbose=40, max_nbytes="1.8G")(delayed(prob_func)(i) for i in range(2))
    
    Traceback (most recent call last):
    
      File "<ipython-input-1-870c0185c434>", line 15, in <module>
        a = Parallel(n_jobs=2, verbose=40, max_nbytes="1.8G")(delayed(prob_func)(i) for i in range(2))
    
      File "/Users/Nick/anaconda/lib/python3.6/site-packages/joblib/parallel.py", line 789, in __call__
        self.retrieve()
    
      File "/Users/Nick/anaconda/lib/python3.6/site-packages/joblib/parallel.py", line 699, in retrieve
        self._output.extend(job.get(timeout=self.timeout))
    
      File "/Users/Nick/anaconda/lib/python3.6/multiprocessing/pool.py", line 608, in get
        raise self._value
    
    MaybeEncodingError: Error sending result: '[0    1
    0    1
    
    [long list of 0    1  omitted...]
    
    0    1
    0    1
    dtype: int64]'. Reason: 'error("'i' format requires -2147483648 <= number <= 2147483647",)'

**Comment by borntospark0** [comment]
%%time

for i in range(1,30,2):
    # instantiate learning model (k = 30)
    knn = KNeighborsClassifier(n_neighbors=i, algorithm='brute',**n_jobs=8**)

    # fitting the model on cross validation train
    knn.fit(X_tr, y_tr)

    # predict the response on the cross validation train
    y_pred_cv = knn.predict(X_cv)

    # evaluate CV accuracy
    acc = accuracy_score(y_cv, y_pred_cv, normalize=True) * float(100)
    print('\nCV accuracy for k = {} is {}'.format(i, acc))
    print("\nAccuracy on cv set: %0.3f%%"%(accuracy_score(y_cv, y_pred_cv)*100))
    print("\nPrecision on cv set: %0.3f"%(precision_score(y_cv, y_pred_cv)))
    print("\nRecall on cv set: %0.3f"%(recall_score(y_cv, y_pred_cv)))
    print("\nF1-Score on cv set: %0.3f"%(f1_score(y_cv, y_pred_cv)))
    print()
    print()

**error is coming as below :pls help in rectifying the error:**

MaybeEncodingError                        Traceback (most recent call last)
<timed exec> in <module>

/usr/local/lib/python3.5/site-packages/sklearn/neighbors/classification.py in predict(self, X)
    143         X = check_array(X, accept_sparse='csr')
    144 
--> 145         neigh_dist, neigh_ind = self.kneighbors(X)
    146 
    147         classes_ = self.classes_

/usr/local/lib/python3.5/site-packages/sklearn/neighbors/base.py in kneighbors(self, X, n_neighbors, return_distance)
    355             if self.effective_metric_ == 'euclidean':
    356                 dist = pairwise_distances(X, self._fit_X, 'euclidean',
--> 357                                           n_jobs=n_jobs, squared=True)
    358             else:
    359                 dist = pairwise_distances(

/usr/local/lib/python3.5/site-packages/sklearn/metrics/pairwise.py in pairwise_distances(X, Y, metric, n_jobs, **kwds)
   1245         func = partial(distance.cdist, metric=metric, **kwds)
   1246 
-> 1247     return _parallel_pairwise(X, Y, func, n_jobs, **kwds)
   1248 
   1249 

/usr/local/lib/python3.5/site-packages/sklearn/metrics/pairwise.py in _parallel_pairwise(X, Y, func, n_jobs, **kwds)
   1094     ret = Parallel(n_jobs=n_jobs, verbose=0)(
   1095         fd(X, Y[s], **kwds)
-> 1096         for s in gen_even_slices(Y.shape[0], n_jobs))
   1097 
   1098     return np.hstack(ret)

/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py in __call__(self, iterable)
    787                 # consumption.
    788                 self._iterating = False
--> 789             self.retrieve()
    790             # Make sure that we get a last message telling us we are done
    791             elapsed_time = time.time() - self._start_time

/usr/local/lib/python3.5/site-packages/sklearn/externals/joblib/parallel.py in retrieve(self)
    697             try:
    698                 if getattr(self._backend, 'supports_timeout', False):
--> 699                     self._output.extend(job.get(timeout=self.timeout))
    700                 else:
    701                     self._output.extend(job.get())

/usr/lib/python3.5/multiprocessing/pool.py in get(self, timeout)
    606             return self._value
    607         else:
--> 608             raise self._value
    609 
    610     def _set(self, i, obj):

MaybeEncodingError: Error sending result: '[array([[1.64749465, 1.55572944, 1.69051779, ..., 1.85987782, 1.66266313,
        1.90668004],
       [1.67146819, 1.78745333, 1.78138775, ..., 1.66264116, 1.58071498,
        1.68958728],
       [1.57148647, 1.54387257, 1.6625694 , ..., 1.79715374, 1.63426997,
        1.76584759],
       ...,
       [1.6138168 , 1.63739606, 1.7464654 , ..., 1.81338028, 1.60118216,
        1.46279678],
       [1.65467385, 1.69937112, 1.68978696, ..., 1.90614982, 1.45027594,
        1.90606051],
       [1.61054637, 1.74764512, 1.74658613, ..., 1.95854167, 1.71440764,
        1.74770114]])]'. Reason: 'error("'i' format requires -2147483648 <= number <= 2147483647",)'
